src/core/src/op/concat.cpp

```diff
@@ -52,34 +52,43 @@ std::shared_ptr<Node> Concat::clone_with_new_inputs(const OutputVector& new_args
     return std::make_shared<Concat>(new_args, m_axis);
 }
 
-bool Concat::evaluate(TensorVector& outputs, const TensorVector& inputs) const {
-    OV_OP_SCOPE(v0_Concat_evaluate);
-    OPENVINO_ASSERT(outputs.size() == 1);
-
+template <typename T>
+void evaluate_concat(const Concat* node, TensorVector& outputs, const TensorVector& inputs) {
     const auto inputs_count = inputs.size();
-    std::vector<const char*> arg_bufs(inputs_count);
     std::vector<Shape> arg_shapes;
     std::vector<PartialShape> input_shapes;
     arg_shapes.reserve(inputs_count);
     input_shapes.reserve(inputs_count);
 
+    std::vector<const T*> arg_bufs(inputs_count);
     auto arg_buf = arg_bufs.begin();
     for (auto& input : inputs) {
-        *arg_buf = static_cast<const char*>(input.data());
+        *arg_buf = static_cast<const T*>(input.data());
         ++arg_buf;
         const auto& input_shape = input.get_shape();
         arg_shapes.emplace_back(input_shape);
         input_shapes.emplace_back(input_shape);
     }
 
-    const auto& out_shape = shape_infer(this, input_shapes).front().to_shape();
+    const auto& out_shape = shape_infer(node, input_shapes).front().to_shape();
     outputs.front().set_shape(out_shape);
     reference::concat(arg_bufs,
-                      static_cast<char*>(outputs.front().data()),
+                      static_cast<T*>(outputs.front().data()),
                       arg_shapes,
                       out_shape,
-                      ov::util::normalize(get_axis(), out_shape.size()),
+                      ov::util::normalize(node->get_axis(), out_shape.size()),
                       outputs.front().get_element_type().size());
+}
+
+bool Concat::evaluate(TensorVector& outputs, const TensorVector& inputs) const {
+    OV_OP_SCOPE(v0_Concat_evaluate);
+    OPENVINO_ASSERT(outputs.size() == 1);
+
+    if (outputs.front().get_element_type() == ov::element::string) {
+        evaluate_concat<std::string>(this, outputs, inputs);
```

Comment:
Instead make the whole evaluate as template (there are some code part which not depends on type) and introduce smaller templates and call them like
```
  auto arg_buff = get_buffers<std::string>(inputs);
   reference::concat(...) // for string
```
  
  and do similar for char* and not duplicate non-template code.

Reply:
To be applied and reviewed within:
- https://github.com/openvinotoolkit/openvino/pull/23048